File: sdks/python/src/echo_agent_sdk/segment_range_values.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class SegmentRange:
    start: int = 0
    end: int = 0

    def __post_init__(self) -> None:
        if (
            isinstance(self.start, bool)
            or not isinstance(self.start, int)
            or isinstance(self.end, bool)
            or not isinstance(self.end, int)
            or self.start < 0
            or self.end < 0
        ):
            raise ValueError("segment range bounds must be non-negative integers")

    def len(self) -> int:
        return max(0, self.end - self.start)

    def is_empty(self) -> bool:
        return self.len() == 0

File: sdks/typescript/src/segment_range_values.ts
/** Message-cache segment range values without owning cache state. */
export class SegmentRange {
  public constructor(public readonly start: bigint = 0n, public readonly end: bigint = 0n) {
    if (start < 0n || end < 0n) throw new RangeError("segment range bounds must be non-negative");
    Object.freeze(this);
  }

  public len(): bigint {
    return this.end > this.start ? this.end - this.start : 0n;
  }

  public isEmpty(): boolean {
    return this.len() === 0n;
  }
}
